include/deal.II/lac/trilinos_tpetra_sparsity_pattern.h

```diff
@@ -0,0 +1,1378 @@
+       * the base class for both const and non-const accessor classes into
+       * sparse matrices.
+       *
+       * Note that this class only allows read access to elements, providing
+       * their row and column number. It does not allow modifying the sparsity
+       * pattern itself.
+       *
+       * @ingroup TrilinosWrappers
+       */
+      template <typename Node =
+                  Tpetra::KokkosClassic::DefaultNode::DefaultNodeType>
+      class Accessor
+      {
+      public:
+        /**
+         * Declare type for container size.
+         */
+        using size_type = dealii::types::signed_global_dof_index;
+
+        /**
+         * Constructor.
+         */
+        Accessor(const SparsityPattern<Node> *sparsity_pattern,
+                 const size_type              row,
+                 const size_type              index);
+
+        /**
+         * Row number of the element represented by this object.
+         */
+        size_type
+        row() const;
+
+        /**
+         * Index in row of the element represented by this object.
+         */
+        size_type
+        index() const;
+
+        /**
+         * Column number of the element represented by this object.
+         */
+        size_type
+        column() const;
+
+        /**
+         * Exception
+         */
+        DeclException0(ExcBeyondEndOfSparsityPattern);
+
+        /**
+         * Exception
+         */
+        DeclException3(ExcAccessToNonlocalRow,
+                       size_type,
+                       size_type,
+                       size_type,
+                       << "You tried to access row " << arg1
+                       << " of a distributed sparsity pattern, "
+                       << " but only rows " << arg2 << " through " << arg3
+                       << " are stored locally and can be accessed.");
+
+      private:
+        /**
+         * The matrix accessed.
+         */
+        mutable SparsityPattern<Node> *sparsity_pattern;
+
+        /**
+         * Current row number.
+         */
+        size_type a_row;
+
+        /**
+         * Current index in row.
+         */
+        size_type a_index;
+
+        /**
+         * Cache where we store the column indices of the present row. This is
+         * necessary, since Trilinos makes access to the elements of its
+         * matrices rather hard, and it is much more efficient to copy all
+         * column entries of a row once when we enter it than repeatedly asking
+         * Trilinos for individual ones. This also makes some sense since it is
+         * likely that we will access them sequentially anyway.
+         *
+         * In order to make copying of iterators/accessor of acceptable
+         * performance, we keep a shared pointer to these entries so that more
+         * than one accessor can access this data if necessary.
+         */
+        Teuchos::RCP<std::vector<dealii::types::signed_global_dof_index>>
+          colnum_cache;
```

Comment:
The question to me is what a user would be most likely to use this information for. 

If it is for Trilinos stuff then the RCP would make things a lot smoother (and consistent to the rest of the class)

If it is only in the deal.II part of the code then a shared_ptr would work as well.

But to my understanding they work pretty similar so I kind of lean towards the RCP to have a single counted pointer type throughout the wrappers

Reply:
Right. If you pass an RCP to Trilinos, it's clear that we should use an RCP. Otherwise, like here, I'd say use a `shared_ptr`. But it is not a particularly important issue.

Reply:
I have to take a look into this, whether `colnum_cache` is handed over to Trilinos. 
But I always choose RCP in the Tpetra Wrappers, to make everything consistent.

Reply:
Since it does not make a lot of difference, I changed `colnum_cache` to be a `std::shared_ptr`